### PR TITLE
Add `get_connection_count` function to `GraphEdit`

### LINKED ISSUE #10942: Add function to get number of port connections in GraphNode
### Describe the project you are working on

Plugin based on GraphEdit

### Describe the problem or limitation you are having in your project

I'm trying to limit the number of connections for an input port in a GraphNode node, so that we can only make one connection to one port, and not several, I tried to implement this using is_node_connected(from_node,from_port,to_node,to_port), but we have to enter extra arguments of the first node, without them the function will not work correctly, especially since I would like to know if the port is connected to anything at all, and not to a specific node

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

There is no function that would determine whether a port is connected to something (true/false), or it would be possible to return the number of connections of a certain GraphNode and one of its ports, using a new function

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

The new function would return the number of connections of one of the ports:
 `connections_count(node: StringName, port: int)`

### If this enhancement will not be used often, can it be worked around with a few lines of script?

At the moment, to find out the number of connections, we will have to use more complex methods, which is difficult especially for beginners, this function would be very useful

### Is there a reason why this should be core and not an add-on in the asset library?

Same as above

### PR DESCRIPTION
**Pull request by JekSun97**
_This is my second attempt at this PR, the first one had problems with Git._

Closes: [https://github.com/godotengine/godot-proposals/issues/10942](https://github.com/godotengine/godot-proposals/issues/10942)

Adds a method to GraphEdit to get the number of connections to a given node and port `get_connection_count(to_node,to_port)`.
With this we can easily limit the number of connections using the condition:
```
if get_connection_count(to_node,to_port)<1:
	connect_node(from_node,from_port,to_node,to_port)
```

Test project:
[graphedit_test_func.zip](https://github.com/user-attachments/files/17877358/graphedit_test_func.zip)

### CODE REVIEW

**Comment by JekSun97** [comment]
If anyone has experience in editing XML documentation, I would appreciate help.

**Comment by KoBeWi** [comment]
You forgot to bind the method.

**Comment by JekSun97** [comment]
> You forgot to bind the method.

Thanks for the comment, I was a bit hasty, fixed

**Comment by JekSun97** [comment]
I generated documentation with --doctool but the function disappeared from GraphEdit.xml, what could be the reason?

**Comment by AThousandShips** [comment]
Did you build the editor before running doctool?

**Comment by JekSun97** [comment]
> Вы создали редактор перед запуском doctool?

Yes, I compiled godot.windows.editor.x86_64.exe in the bin folder, and ran it as bin\godot.windows.editor.x86_64.exe --doctool

**Comment by AThousandShips** [comment]
Then did it vanish or did it get moved to the correct position?

**Comment by JekSun97** [comment]
> Then did it vanish or did it get moved to the correct position?

Function has disappeared from the documentation, I also made sure before generating that the function is in the cpp and h file, and it is linked





Sorry for the large number of screenshots

**Comment by AThousandShips** [comment]
That's the version without binding the method so it won't show up, you need to rebuild with it bound

File: doc/classes/GraphEdit.xml
Comment (AThousandShips) [inline doc/classes/GraphEdit.xml:189]:
```suggestion
				Returns the number of connections from [param from_node] to [param from_port] port.
```

**Comment by JekSun97** [comment]
@AThousandShips Oh my god... Did I press something and ruin everything? 😭

**Comment by AThousandShips** [comment]
You pressed the button to add the changes on GitHub, which creates commits on its own, you will need to squash the changes when you're done

**Comment by JekSun97** [comment]
> You pressed the button to add the changes on GitHub, which creates commits on its own, you will need to squash the changes when you're done

If I type `git reset HEAD~` twice, will it put everything back to normal?

**Comment by AThousandShips** [comment]
You need to pull the changes first, you can just make the changes locally and use `git push --force` otherwise

**Comment by AThousandShips** [comment]
I'll let the GUI team review this but the code looks good

File: doc/classes/GraphEdit.xml
Comment (KoBeWi) [inline doc/classes/GraphEdit.xml:189]:
```suggestion
				Returns the number of connections from [param from_port] of [param from_node].
```

File: scene/gui/graph_edit.cpp
Comment (KoBeWi) [inline scene/gui/graph_edit.cpp:343]: I'm not sure if this condition is correct. E.g. if `to_node` is `p_node` and `from_port` is `p_port` (in another node), it could count a connection to another port.

Comment (JekSun97) [reply scene/gui/graph_edit.cpp:343]: This allows us to check the right ports as well as the left ones, without this only the left or right port is counted

Comment (JekSun97) [reply scene/gui/graph_edit.cpp:343]: The test project did not show any problems, this allows us for example to test one node that is attached to itself, from left to right, which is the correct behavior

File: doc/classes/GraphEdit.xml
Comment (JekSun97) [reply doc/classes/GraphEdit.xml:189]: > 

Since I've had a lot of trouble with commit suggestions in the past, to commit and not create a new commit, should I click Commit suggestion or add suggestion to batch ? 😅

File: scene/gui/graph_edit.cpp
Comment (KoBeWi) [reply scene/gui/graph_edit.cpp:343]: This returns 2 connections to Port 0 in the middle node.
image

Comment (JekSun97) [reply scene/gui/graph_edit.cpp:343]: This is actually happening... I had an idea to create two functions, to check `from_node from_port` and `to_node to_port`, but it seems redundant, what else can I come up with to fix this, any tips?

File: doc/classes/GraphEdit.xml
Comment (KoBeWi) [reply doc/classes/GraphEdit.xml:189]: Adding to batch will just add one commit for all suggestions. You have to manually copy the suggestion if you want to avoid GitHub's commits.

File: scene/gui/graph_edit.cpp
Comment (KoBeWi) [reply scene/gui/graph_edit.cpp:343]: You just have to rearrange your condition, so it always checks the correct port.

**Comment by JekSun97** [comment]
I fixed all the problems, but now Git decided to torture me again...

I typed:
```
git commit --amend
git push --force
```

and got this:
```
fatal: The current branch get_connetion_count_method_graphedit has no upstream b
ranch.
To push the current branch and set the remote as upstream, use

git push --set-upstream origin get_connetion_count_method_graphedit

To have this happen automatically for branches without a tracking
upstream, see 'push.autoSetupRemote' in 'git help config'.
```

I typed this `git push --set-upstream origin get_connetion_count_method_graphedit`

and got this:
```
To https://github.com/JekSun97/godot.git
! [rejected] get_connetion_count_method_graphedit -> get_connetion
_count_method_graphedit (non-fast-forward)
error: failed to push some refs to 'https://github.com/JekSun97/godot.git'
```
What magic can I perform to make Git love me again? 😞

File: scene/gui/graph_edit.cpp
Comment (JekSun97) [reply scene/gui/graph_edit.cpp]: > You just have to rearrange your condition, so it always checks the correct port, i.e. use `from_port` only if connecting from this node and `to_port` if connecting to this node.

The logic for checking the condition has been changed, now it works as expected

**Comment by JekSun97** [comment]
I broke everything... Is it possible to create a new clean PR?

**Comment by AThousandShips** [comment]
I would suggest figuring out how to fix it instead, these kinds of mistakes will happen now and again and it's important to not lose the discussion and feedback in a PR by closing one instead of resolving those issues

In this case you should just revert it by using `git reset --hard f6c96b` and then force pushing after fixing the issues

File: scene/gui/graph_edit.cpp
Comment (AThousandShips) [inline scene/gui/graph_edit.cpp:343]:
```suggestion
		if ((conn->from_node == p_node && conn->from_port == p_port) || (conn->to_node == p_node && conn->to_port == p_port)) {
```

**Comment by JekSun97** [comment]
This PR was like a Git exam for me, but thanks to AThousandShips' tips, I got through it without opening a third PR, thanks!

All recommended edits have been applied, now this PR is DEFINITELY READY 🎉

**Comment by AThousandShips** [comment]
Once this has been approved please also remove the `(Fixed incorrect condition order..)` from the commit message, but that can wait until then

Will approve once someone from the GUI/GraphEdit team gives their approval

**Comment by JekSun97** [comment]
> Once this has been approved please also remove the `(Fixed incorrect condition order..)` from the commit message, but that can wait until then
> 
> Will approve once someone from the GUI/GraphEdit team gives their approval

Can I change this now?, before something weird happens to my Git :D

Should I use git commit --amend git push --force ?

**Comment by JekSun97** [comment]
> > Once this has been approved please also remove the `(Fixed incorrect condition order..)` from the commit message, but that can wait until then
> 
> Can I change this now?, before something weird happens to my Git :D
> 
> Should I use git commit --amend git push --force ?

**Comment by AThousandShips** [comment]
That would be the way yes!

**Comment by JekSun97** [comment]
> Once this has been approved please also remove the `(Fixed incorrect condition order..)` from the commit message

Ready

**Review by KoBeWi — APPROVED** [review]

**Review by AThousandShips — APPROVED** [review]

**Comment by Repiteo** [comment]
Thanks! Congratulations on your first merged contribution! 🎉

**Comment by JekSun97** [comment]
I could replace this method in the future with a more useful one, so that it returns an `Array[Dictionary]` array instead of int, so that we could get a list of nodes connected to a specific port, and accordingly return the number of connections using `Array.size`

**Pull request merged**
